### PR TITLE
Expand gpu-kernel-launch synchronous error checking.

### PR DESCRIPTION
**Pull request by DickJC123**
Background: @zhreshold and @szha discovered that if MXNet were built without a suitable kernel binary or PTX for the GPU in the system, MXNet might produce incorrect results, rather than print an error message and exit.  The problem is that the "no kernel for device" error is synchronous to the kernel launch, and so needs to be identified by a cudaPeekAtLastError() or cudaGetLastError() call executed by the CPU thread launching the kernel.  This is best done right after the triple-chevron kernel-launch statement, and mshadow and much of MXNet already perform this check through the MSHADOW_CUDA_POST_KERNEL_CHECK() macro.

This PR expands use of MSHADOW_CUDA_POST_KERNEL_CHECK() to cover any triple-chevron kernel launches where it is missing, e.g. the generic Kernel::Launch and Kernel::LaunchEx routines.  I have verified the original scenario of an improperly built MXNet now exits with an error message.  I did not add a CI test, as this would have required creating a new test-only operator and hooking it to a kernel that fails at launch.

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by marcoabreu** [comment]
While we're not running C++ tests yet, we plan to do so soon. Wouldn't it be possible to define this test-operator in the test directory, leading to it only being compiled when we're running a test build?

File: tests/python/gpu/test_operator_gpu.py [hunk 1]
```diff
@@ -1478,6 +1479,48 @@ def test_cross_device_autograd():
...
+    c = a / b
+    f = c.bind(mx.gpu(0), { 'a':mx.nd.array([1,2,3],ctx=mx.gpu(0)),
+                            'b':mx.nd.array([],ctx=mx.gpu(0))})
+    f.forward()
+    g = f.outputs[0].asnumpy()
+
+def test_kernel_error_checking():
+    # Execute test in a separate process with its own execution Engine.
+    if not can_fork_exec_process():
+        print('SKIP: needs process fork-exec support.')
```
Comment (marcoabreu) [inline tests/python/gpu/test_operator_gpu.py]: Under what circumstance would this not be supported?

Comment (DickJC123) [reply tests/python/gpu/test_operator_gpu.py]: Python2 linux for one.  I was looking for a test that would be run at least once in the full CI environment.  I believe all Python3 environments should perform the test.

**Comment by marcoabreu** [comment]
What's the problem if an exception is getting thrown in the current process? Would the process terminate without notice and proper error message or why exactly do you prefer creating a new process?

**Comment by DickJC123** [comment]
The main python thread is protected by the try-catch created by the API_BEGIN() and API_END() macros in src/c_api_common.h.  Any exceptions thrown in separate worker threads are not caught by this try-catch, so those exceptions cause terminate() to be called which kills the CI.  One thing I considered was setting up MXNET_ENGINE_TYPE=NaiveEngine which performs all actions on the one python thread.  The problem with this approach is that the Engine is a singleton and the environment variable is only checked when the first test of the CI calls for an engine. Once again, one needs a separate process to get a fresh engine singleton for a particular test.

So I've today finished a bit of polishing on the way to create the separate test process.  I've confirmed that the test runs properly under both Win and Linux Python3 CI environments, and is appropriately skipped for Python2 environments, where the needed fork-exec support is missing.

I'm satisfied with the PR as it stands now, and hope it will be considered for merging.

**Comment by marcoabreu** [comment]
Thanks, your approach seems the perfect way to go!

**Pull request merged**